builtin/game/hud.lua

```diff
@@ -0,0 +1,235 @@
+		local player = core.get_player_by_name(playername)
+		local id = player_hud_ids[elem_name]
+		if player and id then
+			player:hud_remove(id)
+			update_element(player, player_hud_ids, elem_name, player:hud_get_flags())
+		end
+	end
+	return true
+end
+
+local function cleanup_builtin_hud(player)
+	local name = player:get_player_name()
+	if name == "" then
+		return
+	end
+	hud_ids[name] = nil
+end
+
+
+-- Append "update_hud" as late as possible
+-- This ensures that the HUD is hidden when the flags are updated in this callback
+core.register_on_mods_loaded(function()
+	core.register_on_joinplayer(update_hud)
+end)
+core.register_on_leaveplayer(cleanup_builtin_hud)
+core.register_playerevent(player_event_handler)
+
+
+---- Builtin HUD Elements
+
+--- Healthbar
+
+-- Cache setting
+local enable_damage = core.settings:get_bool("enable_damage")
+
+local function scaleToHudMax(player, field)
+	-- Scale "hp" or "breath" to the hud maximum dimensions
+	local current = player["get_" .. field](player)
+	local nominal
+	if field == "hp" then -- HUD is called health but field is hp
+		nominal = registered_elements.health.elem_def.item
+	else
+		nominal = registered_elements[field].elem_def.item
+	end
+	local max_display = math.max(player:get_properties()[field .. "_max"], current)
+	return math.ceil(current / max_display * nominal)
+end
+
+register_builtin_hud_element("health", {
+	elem_def = {
+		type = "statbar",
+		position = {x = 0.5, y = 1},
+		text = "heart.png",
+		text2 = "heart_gone.png",
+		number = core.PLAYER_MAX_HP_DEFAULT,
+		item = core.PLAYER_MAX_HP_DEFAULT,
+		direction = 0,
+		size = {x = 24, y = 24},
+		offset = {x = (-10 * 24) - 25, y = -(48 + 24 + 16)},
+	},
+	show_elem = function(player, flags)
+		return flags.healthbar and enable_damage and player:get_armor_groups().immortal ~= 1
+	end,
+	event = "health_changed",
+	hud_change = function(player, id)
+		player:hud_change(id, "number", scaleToHudMax(player, "hp"))
+	end,
+})
+
+--- Breathbar
+
+register_builtin_hud_element("breath", {
+	elem_def = {
+		type = "statbar",
+		position = {x = 0.5, y = 1},
+		text = "bubble.png",
+		text2 = "bubble_gone.png",
+		number = core.PLAYER_MAX_BREATH_DEFAULT * 2,
+		item = core.PLAYER_MAX_BREATH_DEFAULT * 2,
+		direction = 0,
+		size = {x = 24, y = 24},
+		offset = {x = 25, y= -(48 + 24 + 16)},
+	},
+	show_elem = function(player, flags, id)
+		local breath     = player:get_breath()
+		local breath_max = player:get_properties().breath_max
+		local show_breathbar = flags.breathbar and enable_damage and player:get_armor_groups().immortal ~= 1
+
+		if id then
+			if breath == breath_max then
+				local player_name = player:get_player_name()
+				-- The breathbar stays for some time and then gets removed.
+				core.after(1, function(player_name, id)
+					local player = core.get_player_by_name(player_name)
+					if player then
+						player:hud_remove(id)
+					end
+				end, player_name, id)
+				hud_ids[player_name].breath = nil
```

Comment:
This means that there will be two breathbars if you lose some breath again in this one second

Reply:
Yes it does, as it was before the refactoring.
Should I add a mechanism to not show it multiple times?

Reply:
That would be nice, but since this is a refactoring, adding a comment describing the known bug would be enough too.